external-crates/move/crates/move-compiler/src/diagnostics/mod.rs

```diff
@@ -361,6 +379,62 @@ fn render_diagnostic(
     diag
 }
 
+fn emit_diagnostics_json(
+    writer: &mut dyn WriteColor,
+    mapped_files: &MappedFiles,
+    diags: Diagnostics_,
+) {
+    writeln!(writer, "[").expect("ICE reporting error");
+    let mut seen: HashSet<Diagnostic> = HashSet::new();
+    let mut diags = diags.diagnostics.into_iter();
+    if let Some(diag) = diags.next() {
+        seen.insert(diag.clone());
+        let rendered = render_diagnostic_json(mapped_files, diag);
+        write!(writer, "{}", rendered).expect("ICE reporting error");
+    }
+    for diag in diags {
+        if seen.contains(&diag) {
+            continue;
+        }
+        writeln!(writer, ",").expect("ICE reporting error");
+        seen.insert(diag.clone());
+        let rendered = render_diagnostic_json(mapped_files, diag);
+        write!(writer, "{}", rendered).expect("ICE reporting error");
+    }
+    writeln!(writer, "\n]").expect("ICE reporting error");
+}
+
+fn render_diagnostic_json(mapped_files: &MappedFiles, diag: Diagnostic) -> String {
+    use std::fmt::Write;
+    let mut output = String::new();
+    let Diagnostic {
+        info,
+        primary_label: (ploc, _pmsg),
+        secondary_labels: _,
+        notes: _,
+    } = diag;
+
+    writeln!(&mut output, "    {{").expect("ICE reporting error");
+    let bloc = mapped_files.location(ploc);
+    writeln!(
+        &mut output,
+        "        \"file\": \"{}\",",
+        mapped_files.files.get(bloc.file_id).unwrap().name()
+    )
+    .expect("ICE reporting error");
+    writeln!(&mut output, "        \"line\": {},", bloc.start.line).expect("ICE reporting error");
```

Comment:
not necessary, but should we just spit this out into some struct and rely on serde_json?

Reply:
Excellent idea! Done.